docs/dml-language-reference.md

```diff
@@ -1536,6 +1537,62 @@ Examples:
 | bias_add             |                             | `ones = matrix(1, rows=1, cols=height*width); output = input + matrix(bias %*% ones, rows=1, cols=numChannels*height*width)`                                |
 | bias_multiply        |                             | `ones = matrix(1, rows=1, cols=height*width); output = input * matrix(bias %*% ones, rows=1, cols=numChannels*height*width)`                                |
 
+### Parameter Server Built-in Function
+SystemML supports large-scale machine learning architecture **data-parallel Parameter Server**, introduced by a built-in function **paramserv**. Currently both local multi-threaded and spark distributed backend are supported to execute the **paramserv** function. For example, in order to train a model in local backend with update strategy BSP, 10 epochs, 64 batchsize, 10 workers, **paramserv** function should look like this:
+
+
+    model'=paramserv(model=params, features=X, labels=Y, val_features=X_val, val_labels=Y_val, upd="fun1", agg="fun2", mode="LOCAL", utype="BSP", epochs=10, batchsize=64, k=10, hyperparams=hParams)
+
+
+**Table**: Inputs of paramserv function
+
+Parameters | Description | Type | Mandatory | Options
+-------- | ----------- | ---------- | ---------- | -------
+model | All the parameters (e.g., the weight and bias matrices) | list | yes | 
+features | Training features | matrix | yes 
+labels | Training labels | matrix | yes
+val_features | Validation features | matrix | yes
+val_labels | Validation labels | matrix | yes
+upd | Physical name of gradient calculation function. The format should be "related path:func name". For example, "./mnist_lenet_paramserv_sgd.dml::gradients". | string | yes
+agg | Physical name of gradient aggregation function. The format should be "related path:func name". For example, "./mnist_lenet_paramserv_sgd.dml::aggregation". | string | yes
+mode | Execution backend where the parameter function is executed | string | yes | "LOCAL", "REMOTE_SPARK"
+utype | Update strategy | string | yes | "BSP", "ASP"
+freq | Frequency of parameters updating. When setting to "BATCH", the gradients will be pushed to server for updating the parameters after each mini-batch. When setting to "EPOCH", the gradients will be pushed to server after each epoch. | string | no | "BATCH"(default), "EPOCH"
+epochs | Number of epoch | integer | yes |
+batchsize | Size of mini-batch, if the update frequence is setted to "EPOCH", this parameter will be ignored. | integer | no | 64(default)
+k | Degree of parallelism | integer | no | Number of vcores, otherwise vcores / 2 if using openblas(default)
```

Comment:
Could you please verify that this workaround is still needed for openblas? I think after we fixed the degree of parallelism during recompilation this should be fine now.